packages/xstate-react/src/useSelector.ts

```diff
@@ -39,13 +39,13 @@ export function useSelector<
     const initialSelected = selector(getSnapshot(actor));
     updateSelectedIfChanged(initialSelected);
 
-    const sub = actor.subscribe((emitted) => {
-      const nextSelected = selector(emitted);
+    const sub = actor.subscribe(() => {
+      const nextSelected = selector(getSnapshot(actor));
```

Comment:
this is something that @davidkpiano has wanted as `getSnapshot` is treated as reading the latest known value which could be out of date. In reality, I don't think this matters that much but this was not using this pattern intentionally. We could maybe reconsider this though.

Actor here is close to things like [`useSubscription`](https://github.com/facebook/react/blob/343710c9230d643ec442acef4634e380c1f0bce9/packages/use-subscription/src/useSubscription.js) and [`useMutableSource`](https://github.com/facebook/react/blob/343710c9230d643ec442acef4634e380c1f0bce9/packages/react-reconciler/src/__tests__/useMutableSource-test.internal.js). And as such I think we should try to stick to semantics coined by the React team for those as that looks like a "blessed" way of doing this kind of thing.

The first notable difference between their implementation and this one is that they would keep the `actor` in the internal state. Based on that they would `setState` with the new actor in the render - using "the derived state from props" pattern.

Why this matter? When the `actor` changes, then with this the next **committed** selected state to the screen will be stale (from the previous actor) and the update with the new state will be scheduled right away. So there are 2 takeaways from this:
1. we go through the render->commit cycle twice
2. the user sees the stale value for longer than they should and if the same source state is also displayed somewhere else on the page then what will be displayed here and what will be displayed there won't match. This is known as "tearing"

With the algorithm used by the React team, this doesn't happen as the new committed value will be the correct one right away - at the cost of calling the render function twice but that should be cheaper.

Reply:
I would like to pull the discussion back a bit though:
1) Previously the hook did not even register actor changes, so we're not really breaking on the render -> commited  commit cycle.
2) `getSnapshot()` is defined in the docs as:
> a function that should return the latest emitted value from the actor.
Defaults to attempting to get the actor.state, or returning undefined if that does not exist. Will automatically pull the state from services.

All that this PR attempts to do is bring the code in line with the documentation.

Tangentially: I now noticed that the `getSnapshot()` change needs to happen here too:
https://github.com/davidkpiano/xstate/blob/7d07bfb1ade3137acb1bb6fff33b23d32bfa91a6/packages/xstate-react/src/useActor.ts#L64

--- 

To bring it back to this specific discussion: 
> When the `actor` changes, then with this the next **committed** selected state to the screen will be stale (from the previous actor) and the update with the new state will be scheduled right away. 

If I understand it correctly, to fix this we would need to adapt this code: https://github.com/facebook/react/blob/343710c9230d643ec442acef4634e380c1f0bce9/packages/use-subscription/src/useSubscription.js#L41-L56

and, in our case, instead of `getCurrentValue`, it would be `getSnapshot`, correct?

Reply:
> so we're not really breaking on the render -> commited commit cycle.

Could you rephrase? I'm not sure if I got what you have meant here.

> All that this PR attempts to do is bring the code in line with the documentation.

Can you get any more recent value by calling `getSnapshot` than the one you have just received in the `subscribe` callback? Those values should be really the same so there should be no need to call `getSnapshot` in the callback. Some actors might not expose a property for reading the "snapshot" but they can still emit a value and this is something that won't be handled by the proposed changes.

> If I understand it correctly, to fix this we would need to adapt this code:

Yes. However, I would start with keeping the `actor` and the `value` in the state to see how it goes. I'm not yet sure if we need to keep the `getSnapshot` there.

Reply:
> Could you rephrase? I'm not sure if I got what you have meant here.
I meant, changing the actor flat out did not work before. Introducing a working version with tearing is a big step in the right direction, even if not a perfect solution.

> Can you get any more recent value by calling getSnapshot than the one you have just received in the subscribe callback?
It's not about being more recent. It's about being able to specify a custom `getSnapshot()` implementation, and that implementation being honored. Currently, overwriting `getSnapshot()` in a way that produces different result from the value emitted on the subscription does not work. Now, a reasonable question is, "Should that even work?" I think that it should, but I'm not married to that position.

> Yes. However, I would start with keeping the actor and the value in the state to see how it goes. I'm not yet sure if we need to keep the getSnapshot there.
Should I try to write a `useSubscription()`-style implementation in this PR?

Reply:
> Introducing a working version with tearing is a big step in the right direction, even if not a perfect solution.

Sure, but since we know that the issue is there we can try to fix avoid it altogether - especially that steps towards avoiding it are clear.

> Still, it seems weird to only use getSnapshot() on the initial value, and not on emit events.

I agree and I had similar concerns when this was introduced. But `TEmitted` is used for both the emitted value and the return type of the `getSnapshot` so both really should be the same and I think this is a reasonable assumption. We don't want to support "arbitrary" snapshots.

Reply:
@Andarist @VanTanev Is this resolved?